### PR TITLE
fix(sdk): unblock valid topology. 

### PR DESCRIPTION
**Pull request by JOCSTAA**
**Description of your changes:**
Enables the creation for more complex topologies of task group types that were previously blocked. 

Sample topology that is now valid: 
```
@dsl.pipeline
def my_pipeline(string: str = 'string'):
    with dsl.ParallelFor([1, 2, 3]):
        op1 = comp()
        with dsl.Condition(string == 'string'):
            grandchild = comp(inp=op1.output)
```

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/components/pipeline_task_test.py [hunk 1]
```diff
@@ -352,7 +352,7 @@ def my_pipeline():
                 second_exit_task = print_op(message='Second exit task.')
                 with dsl.ExitHandler(second_exit_task):
                     x = print_op(message='Inside second exit handler.')
-                    x.after(first_exit_task)
+                    x.after(first_print_op)
```
Comment (connor-mccarthy) [inline sdk/python/kfp/components/pipeline_task_test.py:355]: TODO for me: look into why this change was necessary

File: sdk/python/kfp/components/pipeline_task.py [hunk 2]
```diff
@@ -563,8 +563,13 @@ def my_pipeline():
                 task1 = my_component(text='1st task')
                 task2 = my_component(text='2nd task').after(task1)
         """
+        from kfp.components.tasks_group import TasksGroupType
+
         for task in tasks:
-            if task.parent_task_group is not self.parent_task_group:
+            if task.parent_task_group is not self.parent_task_group and task.parent_task_group.group_type in [
+                    TasksGroupType.CONDITION, TasksGroupType.FOR_LOOP,
```
Comment (connor-mccarthy) [inline sdk/python/kfp/components/pipeline_task.py:570]: Can you please add tests for this change?

**Review by connor-mccarthy — CHANGES_REQUESTED** [review]

File: sdk/python/kfp/components/pipeline_task_test.py [hunk 3]
```diff
@@ -411,6 +411,78 @@ def my_pipeline():
...
+
+            @dsl.component
+            def print_op(message: str):
+                print(message)
+
+            @dsl.component
+            def return_list() -> int:
+                return [1, 2, 3]
+
+            @dsl.pipeline(name='pipeline-with-multiple-exit-handlers')
```
Comment (connor-mccarthy) [inline sdk/python/kfp/components/pipeline_task_test.py:426]: nit: update names throughout or remove (`dsl.pipeline` does not need to be called)

Comment (connor-mccarthy) [reply sdk/python/kfp/components/pipeline_task_test.py:426]: Probably should remove for all in this file -- I see that the other test cases have misleading names also

File: sdk/python/kfp/components/pipeline_task.py [hunk 4]
```diff
@@ -563,8 +563,13 @@ def my_pipeline():
...
+                    TasksGroupType.EXIT_HANDLER
+            ]:
                 raise ValueError(
                     f'Cannot use .after() across inner pipelines or DSL control flow features. Tried to set {self.name} after {task.name}, but these tasks do not belong to the same pipeline or are not enclosed in the same control flow content manager.'
```
Comment (connor-mccarthy) [inline sdk/python/kfp/components/pipeline_task.py:574]: We should probably update the error message, since `.after()` can now be used across sub-DAGs

File: sdk/python/kfp/components/pipeline_task.py (hunk 2 above)
Comment (chensun) [inline sdk/python/kfp/components/pipeline_task.py:570]: I don't think the check based on group_type is the right fix here. I think it would still raise error with the following valid topology case, right?
```python
with dsl.Condition(...):
    task1 = component(...)
    with dsl.Condition(...):
        task2 = component(...)
        task2.after(task1)
```

**Review by chensun — CHANGES_REQUESTED** [review]

File: sdk/python/kfp/components/pipeline_task_test.py [hunk 5]
```diff
@@ -411,6 +411,78 @@ def my_pipeline():
...
+                    one = print_op(message='1')
+                    two = print_op(message='2')
+                three = print_op(message='3').after(one)
+
+            with tempfile.TemporaryDirectory() as tempdir:
+                package_path = os.path.join(tempdir, 'pipeline.yaml')
+                compiler.Compiler().compile(
+                    pipeline_func=my_pipeline, package_path=package_path)
+
+    def test_valid_outside_of_condition_topology_permitted(self):
```
Comment (connor-mccarthy) [inline sdk/python/kfp/components/pipeline_task_test.py:462]: Perhaps I'm missing them, but if there are not tests for the other permitted cases, can you add them?

**Review by connor-mccarthy — CHANGES_REQUESTED** [review]

File: sdk/python/kfp/compiler/compiler_utils.py [hunk 6]
```diff
@@ -427,14 +427,29 @@ def get_dependencies(
...
+
+            # Condition check
+            dependent_group = group_name_to_group.get(upstream_groups[0], None)
+            if isinstance(dependent_group, tasks_group.Condition):
+                raise RuntimeError(
+                    f'Task {task.name} cannot dependent on any task inside a Condition that is not a common ancestor of both tasks'
+                )
+
+            # ExitHandler check
             dependent_group = group_name_to_group.get(upstream_groups[0], None)
```
Comment (connor-mccarthy) [inline sdk/python/kfp/compiler/compiler_utils.py:448]: I'm not very familiar with the functions/objects in this part of the codebase. What is the first element of `upstream_groups`?

**Review by connor-mccarthy — CHANGES_REQUESTED** [review]
This PR is getting very close. Thanks, @JOCSTAA!

File: sdk/python/kfp/compiler/compiler_utils.py [hunk 7]
```diff
@@ -427,14 +427,29 @@ def get_dependencies(
...
-            if isinstance(dependent_group,
-                          (tasks_group.ParallelFor, tasks_group.Condition,
-                           tasks_group.ExitHandler)):
+            if isinstance(dependent_group, tasks_group.ExitHandler):
                 raise RuntimeError(
-                    f'Task {task.name} cannot dependent on any task inside'
-                    f' the group: {upstream_groups[0]}.')
+                    f'Task {task.name} cannot dependent on any task inside a Exithandler that is not a common ancestor of both tasks'
```
[the remark this answers is not in the document]
Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/compiler_utils.py:451]: Apologies -- just remembered we discussed offline and this preserves existing behavior.

File: sdk/python/kfp/compiler/compiler_utils.py (hunk 6 above)
Comment (JOCSTAA) [reply sdk/python/kfp/compiler/compiler_utils.py:448]: The first element of upstream group is the first ancestor(condition, parallelfor or exithandler) that is not common between both tasks. So if that ancestor is a condition or exithandler we should fail

File: sdk/python/kfp/compiler/compiler_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (JOCSTAA) [reply sdk/python/kfp/compiler/compiler_utils.py:451]: Hey Connor, so would it be okay to mark this as resolved, since no action is required?

[the remark this answers is not in the document]
Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/compiler_utils.py]: Perhaps @chensun can offer a second opinion. @chensun, WDYT about `task` consuming from `inner_task` in the pipeline above?

[the remark this answers is not in the document]
Comment (chensun) [reply sdk/python/kfp/compiler/compiler_utils.py]: (Sorry for my slow response)

What if `inner_task` failed? Suppose the exit task should handle it gracefully and allow continuing execution of the remaining tasks, then `task` would reference a non-exists output, isn't that just like a conditional case?

We can translate the above example to Python code like:
```python
try:
    inner_task = identity(message=message)
    fail_op(message='Task failed.')
except:
    ...
finally:
    exit_task = identity(message='Exit handler has worked!')

task = identity(message=inner_task.output)
```
Would this makes sense in the Python programming context? Understand Python doesn't block such usage, but it feels to me the last line of `task = ...` could, and probably should be, moved into the `try` block instead. WDTY?

[the remark this answers is not in the document]
Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/compiler_utils.py]: Thanks for mapping this to normal Python -- that's very helpful.

I think the reason for not moving `task` into the `try` block is the same as for normal Python: it might catch exceptions unrelated to the one you're trying to capture. KFP's `ExitHandler` is analogous to a ["bare" `except`](https://www.flake8rules.com/rules/E722.html) in normal Python, so all the more reason to have the minimum required tasks in the `ExitHandler`.

My personal preference is to permit the above syntax. It valid from a control flow standpoint and I don't see much ambiguity for the author.

[the remark this answers is not in the document]
Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/compiler_utils.py]: Either way, I think we should leave the behavior as it currently stands for this PR and address later if we want to :)

[the remark this answers is not in the document]
Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/compiler_utils.py]: After some testing, I see that there's actually more to enabling this behavior than just unblocking it.

The reference to inner task in `identity(message=inner_task.output)` fails to compile, since we don't have logic to register the outputs from `inner_task` to the parent exit handler sub-DAG.

All the more reason to leave as is for this PR.

**Review by connor-mccarthy — CHANGES_REQUESTED** [review]
Thanks, @JOCSTAA! The test cases are very comprehensive and the logic is good. Just a few refactoring suggestions and nitpicks.

File: sdk/python/kfp/compiler/compiler_utils.py [hunk 8]
```diff
@@ -427,14 +427,43 @@ def get_dependencies(
...
+                                    break
+
+                    if exception:
+                        raise RuntimeError(
+                            f'Tasks cannot depend on upstream tasks inside a ParallelFor. Task {task.name} depends on upstream task {upstream_task.name}.'
+                        )
+
+            # Condition check
+            dependent_group = group_name_to_group.get(upstream_groups[0], None)
+            if isinstance(dependent_group, tasks_group.Condition):
```
[the remark this answers is not in the document]
Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/compiler_utils.py:456]: This comment is rendered mostly irrelevant by [this one](https://github.com/kubeflow/pipelines/pull/8416#discussion_r1026805647), except we probably should still comment and test against this particular case.

File: sdk/python/kfp/compiler/compiler_test.py [hunk 9]
```diff
@@ -1490,5 +1484,315 @@ def pipeline_with_input(boolean: bool = False):
...
+            with dsl.ParallelFor([1, 2, 3]):
+                one = print_op(message='1')
+                with dsl.Condition(return_1_task.output == 1):
+                    two = print_op(message='2').after(one)
+
+        with tempfile.TemporaryDirectory() as tempdir:
+            package_path = os.path.join(tempdir, 'pipeline.yaml')
+            compiler.Compiler().compile(
+                pipeline_func=my_pipeline, package_path=package_path)
+
```
[the remark this answers is not in the document]
Comment (JOCSTAA) [reply sdk/python/kfp/compiler/compiler_test.py:1795]: I added a test case that inverts the `dsl.ParallelFor` and `dsl.Condition` order from `test_downstream_in_condition_nested_in_a_for_loop`

However i think that having each of the pipeline that test each case in its own test function would greatly improve readability and debugging ease.

File: sdk/python/kfp/compiler/compiler_test.py [hunk 10]
```diff
@@ -1490,5 +1484,315 @@ def pipeline_with_input(boolean: bool = False):
...
+        def return_1() -> int:
+            return 1
+
+        @dsl.pipeline()
+        def my_pipeline():
+            return_1_task = return_1()
+
+            one = print_op(message='1')
+            two = print_op(message='2')
+            three = print_op(message='3').after(one)
```
[the remark this answers is not in the document]
Comment (JOCSTAA) [reply sdk/python/kfp/compiler/compiler_test.py:1505]: added some implicitly dependent tasks into some test cases

File: sdk/python/kfp/compiler/compiler_utils.py [hunk 11]
```diff
@@ -427,14 +428,42 @@ def get_dependencies(
...
+                    task_group_type = 'a ' + tasks_group.ParallelFor.__name__
+
                 raise RuntimeError(
-                    f'Task {task.name} cannot dependent on any task inside'
-                    f' the group: {upstream_groups[0]}.')
+                    f'Tasks cannot depend on upstream tasks inside {task_group_type} that is not a common ancestor of both tasks. Task {task.name} depends on upstream task {upstream_task.name}.'
+                )
+
+            # ParralelFor Nested Check
+            # if there is a parrallelFor group type in the upstream parents tasks and there also exists a parallelFor in the uncommon_ancestors of downstream: this means a nested for loop exists in the DAG
```
Comment (connor-mccarthy) [inline sdk/python/kfp/compiler/compiler_utils.py:453]: Want to discuss this check a bit more offline to refine my understanding.

**Review by connor-mccarthy — CHANGES_REQUESTED** [review]
Nice work on this, @JOCSTAA!

File: sdk/python/kfp/compiler/compiler_utils.py [hunk 12]
```diff
@@ -427,14 +428,42 @@ def get_dependencies(
...
+            for group in downstream_groups:
+                if isinstance(
+                        group_name_to_group.get(group, None),
+                        tasks_group.ParallelFor):
+                    for parent_task in upstream_parent_tasks:
+                        if isinstance(
+                                group_name_to_group.get(parent_task, None),
+                                tasks_group.ParallelFor):
+                            raise RuntimeError(
+                                f'Downstream task cannot depend on an upstream task while in a nested {tasks_group.ParallelFor.__name__} group. Task {task.name} depends on upstream task {upstream_task.name}, while {group} is nested in {parent_task}'
```
Comment (connor-mccarthy) [inline sdk/python/kfp/compiler/compiler_utils.py:465]: I think this case this handles is actually slightly narrower than the error message suggests. The issue is not that the downstream task cannot be in a nested `ParallelFor` since the following is allowed:

```python
@dsl.pipeline
def p():
    t1 = c1()
    with dsl.ParallelFor([1,2,3]):
        with dsl.ParallelFor([1,2,3]):
            t1 = c1(x=t1.output)
```

What do you think about athe following revised first sentence:

`'Downstream tasks in a nested ParallelFor cannot depend on an upstream task in a shallower ParallelFor.`

Still leaves a bit more precision to be desired, but I think it captures the idea.

Comment (connor-mccarthy) [reply sdk/python/kfp/compiler/compiler_utils.py:465]: nit: also add period to the end of the message for consistency with the above error message

**Review by connor-mccarthy — CHANGES_REQUESTED** [review]
Final nit, now that we know we're going to keep this final check.

File: sdk/python/kfp/compiler/compiler_utils.py [hunk 13]
```diff
@@ -427,14 +428,42 @@ def get_dependencies(
...
+                if isinstance(dependent_group, tasks_group.ExitHandler):
+                    task_group_type = 'an ' + tasks_group.ExitHandler.__name__
+
+                elif isinstance(dependent_group, tasks_group.Condition):
+                    task_group_type = 'a ' + tasks_group.Condition.__name__
+
+                else:
+                    task_group_type = 'a ' + tasks_group.ParallelFor.__name__
+
                 raise RuntimeError(
```
Comment (connor-mccarthy) [inline sdk/python/kfp/compiler/compiler_utils.py:448]: No action required / not your code: I want to note here that I'm not sure if this should use a `RuntimeError`.  This _is_ a runtime error in the sense that it results an ambiguous runtime topology, but it's not a true "runtime" error at the time it's raised.

This file used `RuntimeError` before this PR and, since this PR actually reduces the set of topologies for which this error would be raised, we don't necessarily need to reconsider this in this PR.

Furthermore, an `Exception` is usually used when the error is attributed to user code, whereas an `Error` is usually used when the error is attributed to something else, such as an environment. In this case, this is user code.

For this reason, I think it would make sense for this to be a custom `InvalidTopologyException` or something similar.

Relatedly, some of the `ValueErrors` from `pipeline_task.py` now become `RuntimeErrors` in this PR, so perhaps that is a reason to consider this in the short term.

**Comment by connor-mccarthy** [comment]
/lgtm

Thank you, @JOCSTAA!

**Review by chensun — APPROVED** [review]
